### ISSUE TITLE
Dropdown occurs ArgumentOutOfRangeException

### ISSUE THREAD

**Issue by thinkaboutcsharp** [enisn/Xamarin.Forms.InputKit#31]
I want to use Dropdown, but I got ArgumentOutOfRangeException when choose item from dropdown list. Tap Dropdown control, shown dropdown list, tap item in that list, occurs exception.
I cannot find my bugs. Please tell me solution.

I tried Xamarin.Forms.Picker using same ItemsSource and SelectedItem binding. I succeeded.
I set two items in dropdown list, which item, I got same result.
I tried change type of PickerSelected property `ReactiveProperty<PickerItem>` to normal PickerItem, or PickerData property `ObservableCollection<PickerItem>` to `List<PickerItem>`, same result.
What did I mistake?

XAML:
```xml
        <ContentPage x:Name="pag_SettingsView" Title="SettingsView" Padding="0, 40, 0, 0">
            <StackLayout>
                <input:Dropdown BorderColor="Green" HorizontalOptions="EndAndExpand" ItemsSource="{Binding PickerData}" SelectedItem="{Binding PickerSelected.Value}"/>
            </StackLayout>
        </ContentPage>
```

C#:
```csharp
        public ReactiveProperty<PickerItem> PickerSelected { get; } = new ReactiveProperty<PickerItem>();
        public ObservableCollection<PickerItem> PickerData { get; } = new ObservableCollection<PickerItem>();`

            PickerData.Add(new PickerItem { Title = "abc", Value = 1 });
            PickerData.Add(new PickerItem { Title = "def", Value = 2 });

        public class PickerItem
        {
            public string Title { get; set; }
            public int Value { get; set; }

            public override string ToString()
            {
                return Title;
            }
        }
```
stacktrace:
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
  at System.Collections.Generic.List`1[T].get_Item (System.Int32 index) [0x00009] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/collections/generic/list.cs:181
  at System.Collections.ObjectModel.Collection`1[T].System.Collections.IList.get_Item (System.Int32 index) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/collections/objectmodel/collection.cs:234
  at Plugin.InputKit.Platforms.iOS.MenuEffect+<>c__DisplayClass8_0.<OnPopupRequest>b__0 (UIKit.UIAlertAction action) [0x00025] in <c2c226894a114fd0a53715ac1245e5b9>:0
  at ObjCRuntime.Trampolines+SDActionArity1V175.Invoke (System.IntPtr block, System.IntPtr obj) [0x00016] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/ObjCRuntime/Trampolines.g.cs:23059
  at at (wrapper native-to-managed) ObjCRuntime.Trampolines+SDActionArity1V175.Invoke(intptr,intptr)
  at at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/UIKit/UIApplication.cs:79
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/UIKit/UIApplication.cs:63
  at XamarinStudy.iOS.Application.Main (System.String[] args) [0x00001] in /Users/ryu/OneDrive - hk sar baomin inc/Dev/XamarinStudy/XamarinStudy.iOS/Main.cs:17

**Comment by enisn** [comment]
@thinkaboutcsharp 
Everything looks normal. I'm trying the same situation with same codes with you.

**Comment by thinkaboutcsharp** [comment]
@enisn 
Thank you. I still cannot find solution.
Possibly my environment is something wrong, I pushed full solution at 
https://github.com/thinkaboutcsharp/TacDev/tree/master/XamarinStudy2
I'm trying on iOS simulator. I don't use Android.

Macbook Air (High Sierra)
Visual Studio Community 2017 for Mac 7.6.3
Simulator Version 10.0
Xamarin.Forms.InputKit 2.2.2

**Comment by mxstudios** [comment]
I was getting this error because the Index being passed to the SelectedItem delegate was always the size of the array, which is 1 higher than the last index. NO matter which Item i chose, it came back with a value of 1 higher than the last index. I am not sure if this is your problem, but I fixed by just not using the index. This might have been because of a change I made though.

Line 82 of Dropdown.cs
 ```
private void Menu_Item_Selected(string item, int index)
        {
            try
            {
                SelectedItem = item;     // ItemsSource[index];
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine(ex.ToString());
            }
        }
```
My problem was that the code that creates the actions was also throwing errors that the enumerator was busy or empty, so I had to change the code in the MenuEffect renderer to use a loop instead.

What I couldn't tell from the existing code, is where it was actually incrementing the index that was being passed to the handler. It looks like the variable "i" would always be 0.

I removed this code:
        
```
   int i = 0;
            var _enumerator = Effect.Parent.ItemsSource.GetEnumerator();
            while (_enumerator.MoveNext())
            {
                actionSheetAlert.AddAction(UIAlertAction.Create(_enumerator.Current?.ToString(), UIAlertActionStyle.Default, (action) => Effect.Parent.InvokeItemSelected(_enumerator.Current?.ToString(), i)));
            }

```
and replaced with 

 ```
int i = 0;

            foreach (var item in Effect.Parent.ItemsSource)
            {
                actionSheetAlert.AddAction(UIAlertAction.Create(item.ToString(), UIAlertActionStyle.Default, 
                                                                (action) => Effect.Parent.InvokeItemSelected(item.ToString(), i)));
                i++;
            }
```
For some reason, the index is always came back as the item count, which is one higher than the last index. But I had no need for it, so I just removed the increment, and they all come back as 0 for me.

Don't know if this was helpful, but I still don't know why I was getting enumerator errors, so I just changed it.

**Comment by thinkaboutcsharp** [comment]
Thank you. It appears that iOS has some mystery. In fact, I got new type of error...

At first, I resolved one error. On iPhone Simulator, I need "ObjCRuntime.Class.ThrowOnInitFailure = false;" statement for escape exception "System.Exception: Could not initialize an instance of the type 'UIKit.UIImage': the native 'initWithContentsOfFile:' method returned nil.". I tried adding "36pt arrow_down.png" in iOS project then the exception does not occur.
I thought that arrow_down.png is optionally. Actually, when I write "ThrowOnInitFailure = false" then dropdown is displayed.

Next is problem. At selecting item, I got new error below.
System.InvalidOperationException: Enumeration has either not started or has already finished.
  at System.Collections.Generic.List`1+Enumerator[T].System.Collections.IEnumerator.get_Current () [0x0001d] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/collections/generic/list.cs:1203
  at Plugin.InputKit.Platforms.iOS.MenuEffect+<>c__DisplayClass8_0.<OnPopupRequest>b__0 (UIKit.UIAlertAction action) [0x00000] in <e65f6656247d426eb1caf6d6764e20a9>:0
  at ObjCRuntime.Trampolines+SDActionArity1V175.Invoke (System.IntPtr block, System.IntPtr obj) [0x00016] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/ObjCRuntime/Trampolines.g.cs:23059
  at at (wrapper native-to-managed) ObjCRuntime.Trampolines+SDActionArity1V175.Invoke(intptr,intptr)
  at at (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
  at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/UIKit/UIApplication.cs:79
  at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.14.0.14/src/Xamarin.iOS/UIKit/UIApplication.cs:63
  at XamarinStudy2.iOS.Application.Main (System.String[] args) [0x00001] in /Users/ryu/Dev/XamarinStudy2/XamarinStudy2.iOS/Main.cs:17

On Android Emulator, I succeeded. Problem occurs only on iPhone Simulator. NuGet version is 2.3.0.

**Comment by mxstudios** [comment]
**_System.InvalidOperationException: Enumeration has either not started or has already finished._**

@thinkaboutcsharp , this is the error I fixed by replacing the code in my [comment](https://github.com/enisn/Xamarin.Forms.InputKit/issues/31#issuecomment-419685720) above with the foreach loop.

I have no idea why it was happening so rather than figure it out I just changed it to alternative.

**Comment by thinkaboutcsharp** [comment]
Thank you. Probably nobody knows its reason...(but Xamarin team?)
I will search another way. But maybe I use some controls in this package except Dropdown. I want to release app for iOS.
I hope that problem will be solved someday!

**Comment by thinkaboutcsharp** [comment]
I tracked error. I met one question.
Because of some dependency, I cannot build solution on my environment. So I cannot try debug or do pull request.

I found this.
From MSDN Collection<T>.GetEnumerator()
"Initially, the enumerator is positioned before the first element in the collection. At this position, Current is undefined. Therefore, you must call MoveNext to advance the enumerator to the first element of the collection before reading the value of Current."

In OnPopupRequest method Current is not read before call MoveNext.
I'm worried but only me?
Thank you.

**Comment by thinkaboutcsharp** [comment]
#36 is same reason?

**Comment by mxstudios** [comment]
I am getting same error when I try and build. @enisn? Any Idea?

**Comment by enisn** [comment]
> I want to analize this event, but cannot resolve dependency.
> 
> > Package Plugin.CurrentActivity 2.1.0.4 is not compatible with netstandard2.0 (.NETStandard,Version=v2.0). Package Plugin.CurrentActivity 2.1.0.4 supports: monoandroid44 (MonoAndroid,Version=v4.4)
> 
> How can I build solution? I know this question is not match here, but I need to run correctly.

You need to implement **Plugin.CurrentActivity** into your project first. Then try install InputKit.
You can find from [here ](https://github.com/jamesmontemagno/CurrentActivityPlugin)CurrentActivity Plugin by James Montemagno.

**Comment by thinkaboutcsharp** [comment]
@enisn Thank you. But is not my need. I'll tell you correct, sorry.
I mean that I pulled InputKit source from here(GitHub) and for analyze I tried to build InputKit solution. but I met that dependency error so I cannot do.
My solution can build and run, but causes InvalidOperationException. Then I read InputKit source, I found one question. So I want to try to add line "Current" call. Result, I got build error.

Addition, I pulled and build, I met new dependency error. ConcurrentActivity's error still lives.
`/Users/ryu/Reference/Xamarin.Forms.InputKit/InputKit/InputKit.csproj: Error NU1202: Package Microsoft.NETCore.UniversalWindowsPlatform 6.0.6 is not compatible with netstandard2.0 (.NETStandard,Version=v2.0). Package Microsoft.NETCore.UniversalWindowsPlatform 6.0.6 supports:`
`netcore50 (.NETCore,Version=v5.0`
 `uap10.0.15138 (UAP,Version=v10.0.15138) (NU1202) (InputKit)`

**Comment by thinkaboutcsharp** [comment]
FYI
I still get InvalidOperationException at 2.3.1.

**Comment by thinkaboutcsharp** [comment]
I have searching other products matching my usage, but I don't find. I know that if I want, make myself. That said, here is a solution. What can I do?

**Comment by enisn** [comment]
That should be solved with #88 
Thanks for @damiansikora85
